src/graphapi/find_engine.cpp

```diff
@@ -36,12 +36,163 @@
+    {
+        assert(graph_ptr);
+
+        if(!isIsomorphic(*graph_ptr, getPatternGraph()))
+        {
+            return false;
+        }
+
+        return hasBiasNode(*graph_ptr);
+    }
+
+    std::vector<Engine> getEngines(OpGraph* graph_ptr) const override
+    {
+        assert(graph_ptr);
+        assert(matches(graph_ptr));
+        auto& graph = *graph_ptr;
+
+        auto* conv = dynamic_cast<OperationConvolutionForward*>(
+            graph.findOutNeighByName(graph.getSourceNode(), "OP_CONVOLUTION_FORWARD"));
+        assert(conv);
+
+        std::size_t in_c  = conv->getX()->GetLengths()[1];
+        std::size_t wei_c = conv->getW()->GetLengths()[1];
+
+        if(wei_c == std::size_t{0})
+        {
+            MIOPEN_THROW(
+                miopenStatusBadParm,
+                "invalid weight tensor provided for graph matching ConvBiasResAddActive pattern");
+        }
+        else if(in_c % wei_c != std::size_t{0})
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "invalid group count from input and weight tensor for graph matching "
+                         "ConvBiasResAddActive pattern");
+        }
+
+        int groupCount = in_c / wei_c;
+
+        auto* add1 =
+            dynamic_cast<OperationPointwise*>(graph.findOutNeighByName(conv, "OP_POINTWISE:ADD"));
+        assert(add1);
+
+        auto* add2 =
+            dynamic_cast<OperationPointwise*>(graph.findOutNeighByName(add1, "OP_POINTWISE:ADD"));
+        assert(add2);
+
+        auto* activ = dynamic_cast<OperationPointwise*>(
+            graph.findOutNeighByName(add2, "OP_POINTWISE:RELU_FWD"));
+        assert(activ);
+
+        auto [add, bias] =
+            isBiasNode(add1) ? std::make_pair(add2, add1) : std::make_pair(add1, add2);
+
+        bool addXIsVirtual = add->getX()->isVirtual();
+        assert(addXIsVirtual || add->getB()->isVirtual());
+
+        auto convertToFloat = [](auto&& arg) { return static_cast<float>(arg); };
+        // The virtual tensor for the add is the result of the convolution, and combining the
+        // alpha1's allow for users to specify the alpha on either, or both nodes, and have it be
+        // correct.
+        float alpha1 =
+            conv->getAlpha() * (addXIsVirtual ? std::visit(convertToFloat, add->getAlpha1())
+                                              : std::visit(convertToFloat, add->getAlpha2()));
+        float alpha2 = addXIsVirtual ? std::visit(convertToFloat, add->getAlpha2())
+                                     : std::visit(convertToFloat, add->getAlpha1());
+
+        auto* zTensor = addXIsVirtual ? add->getB() : add->getX();
+
+        bool biasXIsVirtual = bias->getX()->isVirtual();
+        assert(biasXIsVirtual || bias->getB()->isVirtual());
+
+        auto* biasTensor = biasXIsVirtual ? bias->getB() : bias->getX();
+
+        std::shared_ptr<GraphPatternExecutor> exec = ConvBiasResAddActivForwardExecutor::make(
```

Comment:
It's sure tempting to add a wrapper class that encapsulates all this selection. Something like:
```suggestion
class OperationPointwise_OneVirtual   // TODO: better name
{ // forgive the pseudocode and partial code--hopefully it'll do for a quick suggestion
public:
  Tensor *add, *bias;
  float alpha1, alpha2;
  OperationPointwise_OneVirtual(OpGraph* graph_ptr) 
{
  // do sanity checking along the way
  add = // identify the 'add' node, identify its concrete Tensor, and store
  bias = // identify the 'bias' node, identify its concrete Tensor, and store
  // select and store alpha1 and alpha2
}
  float getAlpha1() { return alpha1; }
  float getAlpha2() { return alpha2; }
};
```
Then this function becomes, starting at 146: 
```suggestion
OperationPointwise_OneVirtual oneVirtual{graph_ptr};
std::shared_ptr<GraphPatternExecutor> exec = ConvBiasResAddActivForwardExecutor::make(
            conv->getX(),
            conv->getW(),
            conv->getConvolution(),
            groupCount,
            oneVirtual.add,
            oneVirtual.bias,
            activ->getY(),
            oneVirtual.getAlpha1(),
            oneVirtual.getAalpha2(),
            std::visit(convertToFloat, activ->getAlpha1()));
```

Reply:
Yea that's a good point. There is a lot of virtual checks sprinkled around the pattern checker code, and adding a quick utility class to manage that would be cleaner. 

Thanks!